crates/wasm-mutate/src/mutators/remove_section.rs

```diff
@@ -0,0 +1,151 @@
+
+/// A mutator that removes a section.
+#[derive(Clone, Copy)]
+pub enum RemoveSection {
+    /// Remove a random custom section, even if it contains data.
+    Custom,
+    /// Remove a random empty section.
+    Empty,
+}
+
+// The SectionId is stored as a `u8`. This macro will ensure that all of the `SectionId`s are
+// matched and also takes care of converting the patterns to `u8` for matching.
+macro_rules! match_section_id {
+    (match ($scrutinee:expr) {
+        $($pat:ident => $result:expr,)*
+        _ => $otherwise:expr,
+    }) => {'result: loop {
+        #![allow(unreachable_code, non_upper_case_globals)]
+        $(const $pat: u8 = SectionId::$pat as u8;)*
+        break 'result match $scrutinee {
+            $($pat => $result,)*
+            _ => $otherwise,
+        };
+        // Check exhaustiveness of the SectionId match
+        match SectionId::Type {
+            $(SectionId::$pat => (),)*
+        };
+    }}
+}
+
+fn is_empty_section(section: &wasm_encoder::RawSection) -> bool {
+    use wasmparser::*;
+    match_section_id!(match (section.id) {
+        Custom => Ok(section.data.is_empty()),
+        Type => TypeSectionReader::new(section.data, 0).map(|r| r.get_count() == 0),
+        Import => ImportSectionReader::new(section.data, 0).map(|r| r.get_count() == 0),
+        Function => FunctionSectionReader::new(section.data, 0).map(|r| r.get_count() == 0),
+        Table => FunctionSectionReader::new(section.data, 0).map(|r| r.get_count() == 0),
+        Memory => MemorySectionReader::new(section.data, 0).map(|r| r.get_count() == 0),
+        Global => GlobalSectionReader::new(section.data, 0).map(|r| r.get_count() == 0),
+        Export => ExportSectionReader::new(section.data, 0).map(|r| r.get_count() == 0),
+        Start => Ok(section.data.is_empty()),
+        Element => ElementSectionReader::new(section.data, 0).map(|r| r.get_count() == 0),
+        Code => CodeSectionReader::new(section.data, 0).map(|r| r.get_count() == 0),
+        Data => DataSectionReader::new(section.data, 0).map(|r| r.get_count() == 0),
+        DataCount => Ok(section.data.is_empty()),
+        Tag => TagSectionReader::new(section.data, 0).map(|r| r.get_count() == 0),
+        _ => Ok(section.data.is_empty()),
+    })
+    .unwrap_or(false)
+}
+
+impl Mutator for RemoveSection {
+    fn can_mutate(&self, config: &WasmMutate) -> bool {
+        match self {
+            &Self::Custom => config
+                .info()
+                .raw_sections
+                .iter()
+                .any(|s| s.id == wasm_encoder::SectionId::Custom as u8),
+            &Self::Empty => config
+                .info()
+                .raw_sections
+                .iter()
+                .any(|s| is_empty_section(s)),
+        }
+    }
+
+    fn mutate<'a>(
+        self,
+        config: &'a mut WasmMutate,
+    ) -> Result<Box<dyn Iterator<Item = Result<Module>> + 'a>>
+    where
+        Self: Copy,
+    {
+        let custom_sections = config
```

Comment:
I think with this change this may be better named something like `removal_candidates` perhaps?